common/core/web/input-processor/src/text/inputProcessor.ts

```diff
@@ -171,6 +171,11 @@ namespace com.keyman.text {
               if(pair.p < KEYSTROKE_EPSILON) {
                 break;
               } else if(timer && timer() >= TIMEOUT_THRESHOLD) {
+                // It's always possible that the thread _executing_ our JS got paused,
+                // even if JS itself is single-threaded.
+                if(alternates.length == 0) {
+                  alternates = undefined;
+                }
```

Comment:
The true core of the fix.  This PR's other changes are simply to promote robustness at other levels, just in case.

Reply:
(From our slack discussion) 

* I'm not entirely clear from the PR description what the root cause of the problem that we are fixing?
* How do we get into the situation where alternates.length == 0? 
* Why is a zero length alternates array a problem?
* Are there other code paths where we might need to do the same thing as you are doing here, or is the extra test in languageProcessor.ts that you've made in this PR sufficient to cover all cases? In which case, is this even needed?

Reply:
The root cause:  if the OS pauses the thread evaluating the script between the time we start our timer (when we save `TIMEOUT_THRESHOLD`) and the first check against the timer, the JS thread could be "out of time" when the OS allows it to resume.  So, from the JS thread's perspective... before a single iteration of the loop may occur, with near-zero thread-execution time.

Should _that_ occur, we will have initialized the `alternates` array, but never been able to put anything within it.  Hence, `alternates.length == 0`.

Why would that be a problem?  Because that isn't _possible_ to optimally handle within the lm-layer's worker thread.  Other changes in this PR can at least prevent a crash, but it'll rob the predictive-text engine of all information about the current keystroke, resulting in poorer suggestions.  Preventing it here allows us to at least preserve info about the actually-typed keystroke.

This is the only place where the `.alternates` property is set.  Even then, the extra test there serves as a stop-gap to perform a similar prevention & fix check for any other paths that may exist in the future.  (It's one of those "extra robustness checks" I referred to above.)

While we _could_ just rely on the languageProcessor.ts check... I'd still like to handle it at the most likely 'source' of the error - the OS-interrupted timer.

------

While there _are_ other possible ways to have a zero-length array without timing out early b/c of the OS, they're pretty niche and unlikely.  For the `sil_euro_latin` keyboard, they're also impossible.  (Requires missing rules and/or rules with `BEEP`.)